{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9c3febc-10e6-484a-ac57-f07ee1eaf74c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "np.set_printoptions(legacy = '1.25')\n",
    "\n",
    "import diempy as diem\n",
    "import time\n",
    "\n",
    "import sys\n",
    "import os\n",
    "\n",
    "%matplotlib widget"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffedcbcd-8426-41a5-85bf-44096c2a758d",
   "metadata": {},
   "source": [
    "# Prepare Data for `diem` analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e991155a-ecd2-428c-9898-1c5669d6f394",
   "metadata": {},
   "source": [
    "## Convert vcf file to diem input files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bafcd0c8-8010-4fcc-a244-2ab8bd54d100",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcfFile = './data/iphiclides_example.vcf.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db35e651-363c-419a-92e7-9fae92271448",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('vcf2diem '+vcfFile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc8c5069-c465-4186-9c5e-e4bf37fd171f",
   "metadata": {},
   "source": [
    "## Update Ploidy information & Recombination rates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da3fadd3-e7b3-4118-ad1e-30b164415f59",
   "metadata": {},
   "source": [
    "vcf2diem generates a meta data file that provides information on chromosome lengths, their relative recombination rates, and the ploidy of each chromosome for each individual.  The ploidy information is used during the polarisation step, and the information about relative recombination rates is used when smoothing the data post-polarisation.  \n",
    "\n",
    "vcf2diem outputs a default ploidy of 2 and a relative recombination rate of 1.0. This must be updated by the user according to their own dataset. \n",
    "\n",
    "**Ploidy**\n",
    "\n",
    "By default, vcf2diem assigns a ploidy of 2 across all individuals and chromosomes.  However, for some chromosomes, this will not be the case.  For example, in an XY system, the male is haploid for the X chromosome. In a ZW system like that of this iphiclides dataset, females are haploid for the Z chromosome.  In our example, we provide a file to update the individual ploidies for this specific chromosome.\n",
    "\n",
    "\n",
    "**Relative Recombination Rates**\n",
    "\n",
    "vcf2diem outputs a default *relative* recombination rate of 1.0. This essentially assumes that every chromosome experiences the same amount of recombination, regardless of the length and regardless of whether it is a sex chromosome.\n",
    "\n",
    "This is approximately true for Lepidopterans, like iphiclides, where each chromosome undergoes one recombination event during meiosis.  However, even in this case, we must account for some nuance. \n",
    "\n",
    "In Lepidoptera, recombination occurs normally in males, the homogametic sex (ZZ), but is completely absent in females, the heterogametic sex (ZW).\n",
    "\n",
    "An autosome spends 1/2 of its time in males, so undergoes recombination in 1/2 of its meiotic events. In contrast, the Z chromsome spends 2/3 of it's time in males, so experiences recombination in 2/3 of its meiotic events. Therefore, the relative rate of recombination is higher on the Z chromosome compared to an autosome: (2/3) / (1/2) = 4/3.\n",
    "\n",
    "We can update this information in the meta data file to account for this difference when smoothing. The file 'relative_rec_rates.txt' shows an example of how this is done. Note that only chromosomes with a rate that is not 1 need to be specified.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "620a00a3-d869-427f-ab90-5f6732aaef8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "metaRawFile = './data/iphiclides_example.vcf.gz.diem_meta.bed'\n",
    "\n",
    "ploidyFile = './user_files/iphiclides_ploidy_update.txt'\n",
    "recFile = './user_files/relative_rec_rates.txt'\n",
    "\n",
    "metaCorrected = './data/iphiclides_example.vcf.gz.diem_meta_corrected.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68d0dcd3-626e-402c-9033-2f4aaf748c36",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.update_meta(metaRawFile,metaCorrected,ploidyFilePath=ploidyFile,recFilePath=recFile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "841bfe90-e084-4dd3-8954-d5167493df91",
   "metadata": {},
   "source": [
    "**other ways relative recombination rates may differ among chromosomes**\n",
    "\n",
    "The relative rate of recombination across chromosomes will vary substantially among organisms. For example, in some organisms, recombination occurs at a per-base-pair rate. In that case, longer chromosomes will have a relatively higher total rate of recombination.\n",
    "\n",
    "To account for this, the user could specify a recombination rate that scales with chromosome size.  For example, if chromosome i has length Li:\n",
    "\n",
    "chr 1 --> L1 --> L1/L1 = 1.0\n",
    "\n",
    "chr 2 --> L2 --> L2/L1\n",
    "\n",
    "chr 3 --> L3 --> L3/L1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21c15d1c-a488-4391-83b5-a413eb7401fb",
   "metadata": {},
   "source": [
    "# Reading data into diemPy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0f93caa-05f1-499b-a4c3-f1787fe33872",
   "metadata": {},
   "source": [
    "We now have our inbut bed file and our *corrected* meta data file.  \n",
    "\n",
    "We can read that in as a 'raw' or 'unpolarized' DiemType object.\n",
    "\n",
    "The DiemType is a class which holds the data and provides the core functionality of diem. Refer to documentation for full details.\n",
    "\n",
    "Here, we will simply show how it is used in a standard workflow. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35a0a17d-a960-4cbd-8cba-3b3bde606536",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedInFile = './data/iphiclides_example.vcf.gz.diem_input.bed'\n",
    "metaCorrected = './data/iphiclides_example.vcf.gz.diem_meta_corrected.bed'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "763b2142-1a54-428b-828a-ae0466270a17",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dRaw = diem.read_diem_bed(bedInFile,metaCorrected)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28eb6736-fd8f-4f43-a770-7f1905667c84",
   "metadata": {},
   "source": [
    "# Polarization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "939cc568-8117-4a91-9157-cbbd6c86647b",
   "metadata": {},
   "source": [
    "To polarize the data, we simply call the polarize() function of the DiemType object we created above.\n",
    "\n",
    "Note:\n",
    "- that it defaults to a parallel evaluation using the full number of available cores. We can change the number of cores manually using the 'ncores=' optional argument.  If n=1, it performs a serial polarization of the data.\n",
    "- it returns a separate *copy* of the DiemType object that is now polarized. That is, the raw object is not changed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d05cdde3-37cf-44f1-b830-698a9bde314b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dPol = dRaw.polarize(ncores=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "022cf82a-4e69-48b2-88c6-60cfb91a3363",
   "metadata": {},
   "source": [
    "We can now save this to a bed file which has the same format as the input file but added information from polarization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793c41fc-f05c-4a73-91b3-3eed37ad9c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedOutFile = './output/diem_output.bed'\n",
    "diem.write_polarized_bed(bedInFile,bedOutFile,dPol)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27c2421a-0f11-4a05-80d4-ec8dfae47e4e",
   "metadata": {},
   "source": [
    "We can also save this directly as a diemtype object. \n",
    "\n",
    "It uses less memory, and it is how we will work with the data moving forward.\n",
    "\n",
    "However, it is only useable within the diem environment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5214cdd0-0a9f-495b-ac90-471bf8a08f73",
   "metadata": {},
   "outputs": [],
   "source": [
    "polarizedDiemTypeOutFile = './output/example_polarized.diemtype'\n",
    "\n",
    "diem.save_DiemType(dPol,polarizedDiemTypeOutFile)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6259607d-3276-4085-833e-fa3ce77c00f0",
   "metadata": {},
   "source": [
    "If we have already polarized the data, we can load the data back in either way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc5e6ac4-929f-472b-9382-3e4c08aab9bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19a9750-d95e-428b-90ec-1b1c6ec6a012",
   "metadata": {},
   "outputs": [],
   "source": [
    "dPol = diem.load_DiemType(polarizedDiemTypeOutFile) \n",
    "#or\n",
    "dPol = diem.read_diem_bed(bedOutFile,metaCorrected)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e17d1432-aebd-4b0e-bfe9-e4997d84e59f",
   "metadata": {},
   "source": [
    "# Hybrid Indices and Sorting "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85f0c6dc-5eff-4372-9b3d-aba70ad25b70",
   "metadata": {},
   "source": [
    "the `diem` algorithm sorts individuals by hybrid index and seeks to maximize the differences between two sets of individuals on either side of a barrier to gene flow.\n",
    "\n",
    "although the hybrid indices are calculated on polarization, the individuals are not sorted by hybrid index as we can see here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ee65f0a-5342-4f70-ba35-bdc04743b5bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(figsize = (6,4))\n",
    "ax.plot(dPol.HIs,marker='.',)\n",
    "plt.ylim(0,1)\n",
    "ax.set_ylabel('HI')\n",
    "ax.set_xlabel('individual idx')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bde02d9-2e3c-4df3-b26e-ae3e97787d36",
   "metadata": {},
   "source": [
    "We can use the sort() function to sort the individuals. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02504da1-756a-4d70-8bba-209ae884da43",
   "metadata": {},
   "outputs": [],
   "source": [
    "dPol.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a3b4468-493e-4955-b7fc-e2f7ddf8ed59",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(figsize=(6,4))\n",
    "ax.plot(dPol.HIs,marker='.',)\n",
    "plt.ylim(0,1)\n",
    "ax.set_ylabel('HI')\n",
    "ax.set_xlabel('individual idx')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ca15fd-0507-40e0-adb5-85468a0127c3",
   "metadata": {},
   "source": [
    "# Plotting a Painting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b98c1691-652e-40a6-8476-0a80fb6cfa46",
   "metadata": {},
   "source": [
    "Now that we have polarized and sorted the data, let's have a look at it using the plot_painting() function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9df14933-09d4-4fac-a39b-12e9d91eb870",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plot_painting?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c48704-3486-48fc-9b1f-1e0b73102586",
   "metadata": {},
   "outputs": [],
   "source": [
    "#here we are plotting chromosome 9\n",
    "diem.plot_painting(dPol.DMBC[9],names=dPol.indNames)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a86f3ffd-b727-413e-be1a-145d625462c4",
   "metadata": {},
   "source": [
    "Simply for illustration, let's compare this to the null polarization that diem started with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e544bf64-4446-4244-830d-f131ff2cbcd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plot_painting(diem.diemtype.flip_polarity(dPol.DMBC[9],dPol.initialPolByChr[9]),names=dPol.indNames)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c108d8d1-6d60-49e5-a654-65f0a1fc2e99",
   "metadata": {},
   "source": [
    "# The Diagnostic Index (DI) and Thresholding"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "939cdfab-a788-414d-8c37-a56dd14897d2",
   "metadata": {},
   "source": [
    "## The DI distribution "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fb55c21-bc1f-44fc-b769-9410a810009b",
   "metadata": {},
   "source": [
    "Here we plot the distribution of diagnostic indices across the data.  For each of the final marker configurations, the diagnostic index tells use how well the marker matches does at explaining our data.  The least-negative value is the best.\n",
    "\n",
    "In teal, we plot a histogram of the DI values for each unique marker.\n",
    "\n",
    "In magenta, the counts of each marker are taken into account."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17231f32-f5c1-47cf-ab79-7f39cb73b436",
   "metadata": {},
   "outputs": [],
   "source": [
    "DIValsWithRepeats = np.hstack(dPol.DIByChr)\n",
    "DIValsUnique = np.unique(DIValsWithRepeats)\n",
    "\n",
    "fig,ax = plt.subplots(figsize=(5,4))\n",
    "(counts,bins,patches) = ax.hist(DIValsWithRepeats,bins=500,color='darkmagenta')\n",
    "ax.hist(DIValsUnique,bins=bins,color='lightseagreen')\n",
    "plt.yscale('log')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9ecce6e-f619-4fa7-b289-9f41cbd2527f",
   "metadata": {},
   "source": [
    "We provide a function that will characterize the marker configurations, and provide information about each:  the count, the DI, the number of differences from the 'ideal marker' (the one with the highest DI) and how  many of the individuals have missing data for that marker. It returns a pandas dataframe.\n",
    "\n",
    "This could be useful for exploring the data in greater detail when deciding on a threshold value in later steps.\n",
    "\n",
    "**note** that this could be memory and time intensive to generate if there are many marker types.  We recommend characterizing only a subset of the most-informative markers to be on the safe side.  It is those markers that we will retain through thresholding later.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00fbb496-1607-4ec1-8d9f-69ebb056a966",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfMarkers = diem.characterize_markers(dPol.apply_threshold(-50))\n",
    "dfMarkers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87a19d32-8533-4aba-ab43-b2d3c14bbcfb",
   "metadata": {},
   "source": [
    "## applying the final threshold chosen"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3842f21-6b2a-45d8-9067-4ff29f0bfa92",
   "metadata": {},
   "source": [
    "The DI distribution for every dataset will be unique.  In the **Visualization** section of the tutorial, we provide further tools and guidance on choosing a threshold. \n",
    "\n",
    "For illustration here, we pick a threshold of -30.  \n",
    "\n",
    "Thresholding returns a copy of the DiemType object, and it prints the proportion of markers retained for each chromsoome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58625c4c-0173-469f-9cb1-b25b5b076633",
   "metadata": {},
   "outputs": [],
   "source": [
    "dThresh = dPol.apply_threshold(-30)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9694e73c-ba3f-47aa-a7cd-79957f9bbc60",
   "metadata": {},
   "source": [
    "After thresholding, there will be fewer markers present, and those that remain have configurations which better support the genetic barrier. In effect, the data looks cleaner. We can see this by plotting chromosome paintings with markers in their physical positions, below, both before and after thresholding. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d2d1c01-a64b-436c-8f25-cd82bc025cbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plot_painting_with_positions(dPol.DMBC[9],dPol.posByChr[9],names=dPol.indNames)\n",
    "diem.plot_painting_with_positions(dThresh.DMBC[9],dThresh.posByChr[9],names=dPol.indNames)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "157827df-3d7e-4eb5-9551-2189e1b97ee4",
   "metadata": {},
   "source": [
    "Below, we compare the hybrid indices before (blue) and after (red) thresholding: \n",
    "\n",
    "We can see that the separation of individuals is clearer now, and their hybrid indices have also changed.\n",
    "\n",
    "We will not do this for now, but we could resort the individuals by HI by calling *dThresh.sort()*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9c0eaca-82a3-406e-be47-97b31357be32",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(figsize = (6,4))\n",
    "ax.plot(dPol.HIs,marker='.',color='b')\n",
    "ax.plot(dThresh.HIs,marker='.',color='r')\n",
    "ax.set_ylim(0,1)\n",
    "ax.set_ylabel('HI')\n",
    "ax.set_xlabel('individual idx')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f20ba98-4393-490b-a92a-67667288ebb1",
   "metadata": {},
   "source": [
    "# Smoothing the Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e285eea4-8574-4d19-9f54-e1a11da7360c",
   "metadata": {},
   "source": [
    "After thresholding, we will see much clearer tracts representing shared ancestry with respect to the barrier.  However, there will be some bits of missing data, mis-called sites, etc. that will break up the tracts. We want to smooth this over for later analyses. We also care about tracts generated by cross-over recombination, but gene conversion events could also cause short changes in state and break up our tracts.  Therefore, we want to smooth over this noise and this signal in the data.\n",
    "\n",
    "In the manuscript **(insert link when available)**, we explain the Laplacian smoothing kernel and how this smoothing is done on a recombination-based (map) distance rather than physical base pair position.\n",
    "\n",
    "The default behaviour of DiemPy is to normalize the physical distances by chromosome length and use a global smoothing scale parameter across all chromosomes. This assumption can be relaxed, and we discuss this later in the **Advanced Smoothing** section.\n",
    "\n",
    "\n",
    "Here, we use the default global scale, and to pick a smoothing scale, we examine how the data changes for different values.  Below, we calculate the number of sites that change across the whole dataset using the laplacian smoothign kernel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f09ff31e-aed0-4871-939b-910d3c530b91",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "scalesToTry = [1e-8,5e-8,1e-7,5e-7,1e-6,5e-6,1e-5,5e-5,1e-4,5e-4]\n",
    "sitesDiffByScale = []\n",
    "for scale in scalesToTry:\n",
    "    dSmoothedTest = dThresh.smooth(scale)\n",
    "    kdiffs = diem.count_site_differences(dSmoothedTest.DMBC,dThresh.DMBC)\n",
    "    sitesDiffByScale.append(kdiffs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "704c66ce-91c4-41ff-a4ca-6a3d278ea412",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(figsize=(6,4))\n",
    "ax.plot(scalesToTry,sitesDiffByScale,marker='.')\n",
    "ax.set_ylabel('# of sites changed')\n",
    "ax.set_xlabel('Laplace scale')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff30b64e-aa9c-4df3-ac0a-2b898c2b1e2f",
   "metadata": {},
   "source": [
    "We see that by around scale = 5e-5, the single-site changes and short tracts have been smoothed over, and we choose this as our final smoothing scale.\n",
    "\n",
    "**note** smoothing returns a copy of the dataset with the same order (i.e. HIs are not recomputed and data is not resorted). That way, comparisons can be made easily before and after smoothing. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ee8f3c8-162d-418f-9147-ecc94dc3422f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dSmoothed = dThresh.smooth(5e-5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57571d6e-50bb-4687-9d49-451750571878",
   "metadata": {},
   "source": [
    "We can compare before and after smoothing to see how it has affected our data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24bba78b-b0e6-4717-8e51-b2cf294ba537",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plot_painting(dThresh.DMBC[9],dThresh.indNames)\n",
    "diem.plot_painting(dSmoothed.DMBC[9],dSmoothed.indNames)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6f0d690-66b2-4771-ba6e-9e9a1c75b9d6",
   "metadata": {},
   "source": [
    "We can also see how smoothing has affected our HIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6682fbe1-f65d-4e50-9a72-e6f74c9ea1f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(figsize=(6,4))\n",
    "ax.plot(dThresh.HIs,marker='.',color='b')\n",
    "ax.plot(dSmoothed.computeHIs(),marker='.',color='r')\n",
    "ax.set_ylabel('HI')\n",
    "ax.set_xlabel('individual idx')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14885767-59ac-4e38-ad30-e9dc3d8c4872",
   "metadata": {},
   "source": [
    "After smoothing, we can see that the hybrid indices have changed, and we now do a final sort of the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4df4a02d-6e9b-4feb-8b15-e0a71721c3af",
   "metadata": {},
   "outputs": [],
   "source": [
    "dSmoothed.sort()\n",
    "diem.plot_painting(dSmoothed.DMBC[9],dSmoothed.indNames)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "468f85ae-787e-43dd-a8e6-f68aa7ca9fb8",
   "metadata": {},
   "source": [
    "# Constructing Contigs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd6d0853-3563-4b28-bce0-394f0516cdfe",
   "metadata": {},
   "source": [
    "Downstream analyses of diem output may focus on the lengths of tracts of shared ancestry across the gene flow barrier.\n",
    "\n",
    "Here, we provide the function create_contig_matrix() that generates 'contigs' for each chromosome of each individual.\n",
    "\n",
    "a 'contig' is a summary of the data.  It is the sequence of contiguous intervals that form a givien chromosome \n",
    "\n",
    "Once we have created the contigs, we can export per-individual contig files in bed formatting "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7806b16-d8f5-40eb-8de6-f5ea9d0c94a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dSmoothed.sort() # ensure that dSmoothed is sorted.\n",
    "dSmoothed.create_contig_matrix()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dfd71fd-edd5-43dc-bf26-9b560e47e4e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "contingOutputDirectory = './output/contigs/'\n",
    "\n",
    "diem.export_contigs_to_ind_bed_files(dSmoothed,contingOutputDirectory)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "192bb620-51f5-485a-a74b-19e63e1865ea",
   "metadata": {},
   "source": [
    "# Using Contigs & intervals"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50367fc1-9d68-44af-90cf-7be4ed24c513",
   "metadata": {},
   "source": [
    "Within the diemtype objects, contigs are stored in the *contigMatrix*\n",
    "\n",
    "the *contigMatrix* is an array wit shape nChrs x nIndividuals. and we can use splicing to access specific contigs. \n",
    "\n",
    "For chromosome indexed 1 and individual indexed 8:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c19d201a-556a-4809-ae6d-8162473f83a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleContig = dSmoothed.contigMatrix[1,8]\n",
    "exampleContig.__dict__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25381047-eac6-4077-8d9a-36e8f576ca21",
   "metadata": {},
   "source": [
    "A contig stores the individual and chromosome names to which it belongs as well as a list of intervals and the number of intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29bd8323-1863-4327-9f39-94d9c027eb7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "?exampleContig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fb733a6-55b4-4cbf-a2a2-c7c2b0639778",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleContig.num_intervals"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95a7b300-f1a2-45e6-974d-6ca6cdf47e15",
   "metadata": {},
   "source": [
    "an **interval** is a contiguous segment of genome with the same ancestry.\n",
    "\n",
    "It stores the chromosome and individual to which it belongs. the indices and positions of the left- and right-most markers,\n",
    "and the state of the individual: 0, 1, 2, or 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3036113a-61d3-404b-9f56-5ce1680aeedb",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleInterval = exampleContig.intervals[0]\n",
    "?exampleInterval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b2fc74d-c511-40a2-899f-21e82770dd75",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleInterval.__dict__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b567b6a3-8d99-45c0-aaa2-33bd26c06278",
   "metadata": {},
   "source": [
    "We can get the interval's span easily:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b706378-6734-4bf7-9458-e1c2210cd663",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleInterval.span()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53d0f834-fcbb-485d-adb3-48fed5a6594b",
   "metadata": {},
   "source": [
    "And we can get the intervals map span (on a unit scale) if we provide the length of the chromsome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2ebba57-cbed-430f-97dc-fafca76a2c34",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleInterval.mapSpan(dSmoothed.chrLengths[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ea5e546-cf8e-4693-be87-e09446d265d2",
   "metadata": {},
   "source": [
    "Here is the information about all the intervals that form the genotype of our example contig:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63deb48d-aeb6-4396-b149-14888eddd053",
   "metadata": {},
   "outputs": [],
   "source": [
    "[x.__dict__ for x in exampleContig.intervals]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78ca23ff-fab9-4f84-bd3e-f2ae1b243703",
   "metadata": {},
   "source": [
    "We can plot the tract length distribution for all intervals with state '1'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38727955-362e-445a-a6b8-21fbf880d0e4",
   "metadata": {},
   "source": [
    "# Plotting tract length distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83af4ef1-26dc-4621-9c53-7509a188472e",
   "metadata": {},
   "outputs": [],
   "source": [
    "oneTracts = dSmoothed.get_intervals_of_state(1)\n",
    "oneLengths = [x.span() for x in oneTracts]\n",
    "fig,ax = plt.subplots()\n",
    "ax.hist(oneLengths,color='m')\n",
    "ax.set_ylabel('counts')\n",
    "ax.set_xlabel('tract length')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96f8aafe-c3ce-4993-8534-26dcebc6110f",
   "metadata": {},
   "source": [
    "We can plot the distribution of the different types of tracts together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4858008-3413-4ada-b2c3-a19e46ace485",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots()\n",
    "ax.hist([x.span() for x in dSmoothed.get_intervals_of_state(1)],bins = 50,color='m',alpha = 0.5)\n",
    "ax.hist([x.span() for x in dSmoothed.get_intervals_of_state(2)],bins=50,color='y', alpha = 0.5)\n",
    "ax.hist([x.span() for x in dSmoothed.get_intervals_of_state(3)],bins=50,color='teal', alpha = 0.5)\n",
    "ax.set_ylabel('counts')\n",
    "ax.set_xlabel('tract length')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ec03da8-d603-40d3-ae09-1fcf93ab244a",
   "metadata": {},
   "source": [
    "# Advanced tract length distribution plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6069df15-bfb6-41a2-9025-9dd15ee94957",
   "metadata": {},
   "source": [
    "We can also subset individuals and chromosomes when getting intervavls of those types."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0b0881f-e992-4775-a45a-743b2b9468d4",
   "metadata": {},
   "source": [
    "Suppose we want to compare the distribution of tract lengths on opposite sides of the barrier. By looking at the HIs of the smoothed data, we see there is one highly admixed individual, IF_1322, so we decide to exclude it here. \n",
    "\n",
    "We then identify the individuals clearly to the 'left (bottom)' and 'right (top)' sides of the barrier. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c9e56f4-d84c-4c24-a2e9-00c77308dabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "[[i,dSmoothed.indNames[i],dSmoothed.HIs[i]] for i in range(len(dSmoothed.indNames))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b2f4533-5832-4125-8a69-56b7231fb755",
   "metadata": {},
   "outputs": [],
   "source": [
    "leftIndividualIndices = np.arange(8)\n",
    "rightIndividualIndices = np.arange(9,20)\n",
    "print(leftIndividualIndices)\n",
    "print(rightIndividualIndices)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "253f5779-5d3c-47a5-a126-5d2b8def768c",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "We can plot the spans of tracts of type '1' for individuals on the 'left' (blue) and 'right' (red) sides of the barrier, respectively. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10668415-f855-45d7-8f55-171509722a04",
   "metadata": {},
   "outputs": [],
   "source": [
    "state1SpansLeftSide = [x.span() for x in dSmoothed.get_intervals_of_state(1,individualSubset=leftIndividualIndices)]\n",
    "state1SpansRightSide = [x.span() for x in dSmoothed.get_intervals_of_state(1,individualSubset=rightIndividualIndices)]\n",
    "\n",
    "fig,ax = plt.subplots()\n",
    "longestSpan = max(max(state1SpansLeftSide),max(state1SpansRightSide))\n",
    "ax.hist(state1SpansLeftSide,bins = 100,range=(0,longestSpan),color = 'b',alpha = 0.5)\n",
    "ax.hist(state1SpansRightSide, bins = 100,range=(0,longestSpan), color = 'r',alpha = 0.5)\n",
    "ax.set_ylabel('counts')\n",
    "ax.set_xlabel('tract length')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12cd1727-8aeb-4dee-a89a-d11c0632c220",
   "metadata": {},
   "source": [
    "Similarly, we can plot the spans of type '3' comparing the left and right sides of the barrier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cc95ba0-dd05-4b0a-ad63-f739243dc3ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "state3SpansLeftSide = [x.span() for x in dSmoothed.get_intervals_of_state(3,individualSubset=leftIndividualIndices)]\n",
    "state3SpansRightSide = [x.span() for x in dSmoothed.get_intervals_of_state(3,individualSubset=rightIndividualIndices)]\n",
    "\n",
    "fig,ax = plt.subplots()\n",
    "longestSpan = max(max(state3SpansLeftSide),max(state3SpansRightSide))\n",
    "ax.hist(state3SpansLeftSide,bins = 100,range=(0,longestSpan),color = 'b',alpha = 0.5)\n",
    "ax.hist(state3SpansRightSide, bins = 100,range=(0,longestSpan), color = 'r',alpha = 0.5)\n",
    "ax.set_xlabel('tract length')\n",
    "ax.set_ylabel('counts')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fba82e6-7c79-4ab2-a69f-3a9af8473c32",
   "metadata": {},
   "source": [
    "# Visualisation"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "23c8d0d3-5977-469f-b89b-ec00f0dcf80f",
   "metadata": {},
   "source": [
    "Polarisation has not changed your data, only coloured it and measured it. The **GenomeSummaryPlot** shows these colours, and the effects of ignoring sites with low diagnostic index (DI).\n",
    "\n",
    "Sort individuals by their hybrid index using the red button. Slide the DI slider to the right to focus on barrier relevant sites. If your data is large, you may need to _click_ at your desired DI value rather than sliding. Mouse-over a plot symbol to see which individual it refers to. You can save the plot using the click-on tool or keyboard shortcut S.\n",
    "\n",
    "**TOUCHDOWN DI**: As you slide DI to the right, first Unpolarisable (U, grey line) data will be filtered out (blanks and genotypes involving third states). Next the span of the hybrid indices (red line) should start to increase (you may wish to hit the red button to re-sort by the new HI). One way of choosing a DI threshold is to move right until 'touchdown' where the hybrid index goes to zero on one side of the barrier, and 1 on the other. An outlier individual may touchdown first, way before others; in that case keep going until other have joined. There is a trade-off: the further to the right, the cleaner the retained data with respect to the barrier, but the fewer sites in the genome are being considered."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d31c8c68-9d5f-4b97-904b-43066cf4e2a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plots.GenomeSummaryPlot(dPol)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c6698559-4b6a-49f2-a5dc-ab560209920e",
   "metadata": {},
   "source": [
    "The effects of DI sliding may not be equal across chromosomes - sex chromosome are often more relevant to barriers. **GenomeMultiSummaryPlot** allows you to explore this by specifying a list of chromosome to summarise as though each is a genome, while still showing the global HI plot (cyan) over all chromosome as reference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6413c400-396d-4130-b9ed-9da6317f9365",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plots.GenomeMultiSummaryPlot(dPol,[0,1,2])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "277fa795-641e-494c-84aa-7659b47e2af4",
   "metadata": {},
   "source": [
    "There are many ways to summarise genomes with respect to the sites diagnosing a barrier. The **GenomicDeFinetti** diagram uses HOM1 HET HOM2 frequencies to place genomes in a ternary plot. The colours are a blend of not just these proportions, but also the Unpolarisable proportion (as White). Skim-sequenced genomes will, for example, have lighter shades due to more missing data.\n",
    "\n",
    "The background parabola is analogous to Hardy-Weinberg equilibrium for HOM1 HET HOM2 frequencies at a single site of a genome. In current context the parabola is the genomic expectation if admixture is proceeding with local HWE in a simlar fashion throughout the genome (ie with similar admixture proportions)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ca73f77-2a35-4c6e-b26e-3f293fbfff6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plots.GenomicDeFinettiPlot(dPol)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0feb8bd6-8c4d-4c79-b6d1-7b7a55741a7c",
   "metadata": {},
   "source": [
    "Again, the effects of DI sliding may not be equal across chromosomes or genome regions. **GenomicMultiDeFinettiPlot** allows you to explore this at the level of chromosomes from the ternary perspective by specifying a list of chromosome to summarise as though each is a genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db1bb1af-4d43-4742-98cb-85e7741a7c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plots.GenomicMultiDeFinettiPlot(dPol,range(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa63346b-329d-4ecf-ba57-6a55cecad915",
   "metadata": {},
   "source": [
    "Another way of exploring the heterogeneity of barrier effects across the genome is to simply look at howmany sites are retained on each chromosome after DI sliding. **GenomicContributionsPlot** shows these contributions, and breaks them down by state: often sex chromsomes will contribute a lot, but with a lower HET proportion.\n",
    "\n",
    "The resemblance to a graphical equaliser from an old stereo raises an analogy: sliding DI to the right, we see the channels (chromosomes) on which barrier signal is strongest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b85e68ba-8902-47aa-bb4d-f83795aced01",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plots.GenomicContributionsPlot(dPol)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fa927f8-9010-408d-9e04-570a6e57a734",
   "metadata": {},
   "source": [
    "It may be that admixture has produced an explosion of genome-types, such that genomes are all unique and their summaries change smoothly across the barrier. Alternatively, for eg a taxon of hybrid origin, a single along-genome blockwise admixture pattern may occur in multiple individuals. **diemPairsPlot** allows you to explore these possibilities through all pairwise comparison of the painted colours of the polarised genomes. \n",
    "\n",
    "The is a genomic O(n<sup>2</sup>) calculation... so choose one DI threshold at a time (maybe a touchdown DI) it may take a while before it plots.... be patient.\n",
    "\n",
    "In the plot genomes are ordered simply by their hybrid index.\n",
    "\n",
    "Blue squares indicate clusters of genomes which share detailed purple/teal long-genome pattern.\n",
    "\n",
    "You may note the diagonal of self-to-self comparison is not always zero. The same distance function is applied for all comparisons (even self-to-self) in order to make it clear that, because _diem_ is ignorant of phase, comparison of HET genome tracts to HET genome tracts do **not** give distance zero. HOM1 vs HOM2 is distance 2. HET vs HET is (either 0 or 2 with phase unknown) => distance 1. (See Petruezela et al Mol Ecol 2025)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a4e767b-5c01-4b5a-9900-f5412fd8e81f",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plots.diemPairsPlot(dPol,-100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa9f2f19-2292-477a-8818-69a2aae96edf",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plots.diemMultiPairsPlot(dPol,[0,1,2,3,4,5,23],-100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3933870b-ee1f-4479-958a-64be4c4c49c2",
   "metadata": {},
   "source": [
    "We mentioned 'detailed purple/teal long-genome pattern'. What does this look like? **diemIrisFromPlotPrep** allows us to see this detail for the whole genome, for all the genomes, all at once. \n",
    "\n",
    "Again this is a computationally intensive task, so first we run a prepatory step **diemPlotPrepFromBedMeta**, which loads the data you saved in the **Polarisation** section and prepares is for the big plot of everything...\n",
    "\n",
    "Give your plot a (short) theme.\n",
    "Set the correct files from **Polarisation**\n",
    "Choose a (touchdown) DI threshold\n",
    "Choose a number of 'pixels' (in multiple of 360... the plot is going to be a ring)\n",
    "Choose the interval between ticks in bases\n",
    "(We will come to smoothing later)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "774564e4-9e9e-4d6e-9bd7-2639468143f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "MyFirstPlotPrep = diem.plots.diemPlotPrepFromBedMeta(\n",
    "    plot_theme=\"tutorial example\",\n",
    "    bed_file_path=bedOutFile,\n",
    "    meta_file_path=metaCorrected,\n",
    "    di_threshold=-57,  \n",
    "    genome_pixels=360*1,\n",
    "    ticks=10000, # bases\n",
    "    smooth=5e-5\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cba54550-cafa-45a2-89b8-7c2da10b7f28",
   "metadata": {},
   "source": [
    "Now we construct an Iris Plot from the prepared data:\n",
    "\n",
    "Mouse-over and you genome/indidivual location will be displayed below the plot... browse your admixture features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd261f7f-5ebf-41a1-9473-1bd756c044c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plots.diemIrisFromPlotPrep(MyFirstPlotPrep)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d6883bb-0def-4d12-9833-2c8a570658e8",
   "metadata": {},
   "source": [
    "We can straighten out this curl to make a rectangular plot using **diemLongFromPlotPrep**:\n",
    "\n",
    "You can choose which chromosomes to include.\n",
    "\n",
    "Again this is browsable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e98759ea-0a73-4b5b-9f92-f8dd37321981",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plots.diemLongFromPlotPrep(MyFirstPlotPrep)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b325c15c-51a8-4ef1-9b9b-bd7716a22ae6",
   "metadata": {},
   "source": [
    "You can limit the plot to a subset of chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34e26516-c157-4552-aa11-e85fbca64e6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plots.diemLongFromPlotPrep(MyFirstPlotPrep,range(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab8aa8bc-d97e-42b3-a099-4825a3f14daf",
   "metadata": {},
   "source": [
    "Or you can stack plots of each of your chromosomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1d0948d-7148-4349-9e53-67ba1eb4cb4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(3):\n",
    "    diem.plots.diemLongFromPlotPrep(MyFirstPlotPrep,[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8718b1c4-73b9-4fc8-927d-733b84de1c1b",
   "metadata": {},
   "source": [
    "# Advanced Polarizing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e299f9fb-15d4-48fc-9503-34fd21da545f",
   "metadata": {},
   "source": [
    "When running diem, we may want to mask some individual and sites.  Here, masked sites/individuals do not *influence* the barrier (i.e. they don't contribute to the likelihood function). However, they *are* polarized during the EM step along with other sites and individuals. \n",
    "\n",
    "We have included two example exclusion files under the user_files subdirectory.\n",
    "\n",
    "Here, we mask individuals that were sampled outside of france and spain.\n",
    "\n",
    "We also (arbitrarily) mask chromosome 2 for this example.\n",
    "\n",
    "Although we have masked both sites and individuals in this example, it possible to specify only one or the other. Any masks have been added to the raw data will be used in polarization. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35002a37-ff64-432b-8bf7-723ef429b147",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #### add information about individuals that do not influence polarization. We say 'excluded' but really mean 'masked'\n",
    "sitesMaskedFile = './user_files/sites_masked.bed'\n",
    "individualsMaskedFile = './user_files/individuals_masked.txt'\n",
    "\n",
    "dRawWithMasks = dRaw.copy()\n",
    "\n",
    "dRawWithMasks.add_individual_exclusions(individualsMaskedFile)\n",
    "dRawWithMasks.add_site_exclusions(sitesMaskedFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "513c672a-f33a-4f09-9105-e4bb13fcefad",
   "metadata": {},
   "outputs": [],
   "source": [
    "dRawWithMasks.siteExclusionsByChr[12]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e01c7640-53bb-41e2-b55d-9aa60d31687f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dPolWithMasks = dRawWithMasks.polarize()\n",
    "dPolWithMasks.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c5ca746-198a-4762-ae7b-96d0e64d3142",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plot_painting(dPolWithMasks.DMBC[9])\n",
    "\n",
    "fig,ax = plt.subplots(figsize=(6,4))\n",
    "ax.plot(dPolWithMasks.HIs,marker='.',color='b')\n",
    "ax.set_ylabel('HI')\n",
    "ax.set_xlabel('individual idx')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e26864e4-cf2f-4e1a-af8d-383e6842d49f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
